core/src/HPX/Kokkos_HPX_ChunkedRoundRobinExecutor.hpp

```diff
@@ -0,0 +1,211 @@
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_HPX_CHUNKEDROUNDROBINEXECUTOR_HPP
+#define KOKKOS_HPX_CHUNKEDROUNDROBINEXECUTOR_HPP
+
+#include <hpx/config.hpp>
+#include <hpx/async_launch_policy_dispatch.hpp>
+#include <hpx/lcos/local/latch.hpp>
+#include <hpx/parallel/executors/execution.hpp>
+#include <hpx/parallel/executors/post_policy_dispatch.hpp>
+#include <hpx/runtime/get_os_thread_count.hpp>
+#include <hpx/runtime/threads/thread_helpers.hpp>
+#include <hpx/traits/is_executor.hpp>
+#include <hpx/traits/is_launch_policy.hpp>
+#include <hpx/util/deferred_call.hpp>
+
+#include <cstddef>
+#include <functional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+namespace Kokkos {
+namespace Impl {
+
+///////////////////////////////////////////////////////////////////////////
+/// A \a ChunkedRoundRobinExecutor creates groups of parallel execution
+/// agents which execute in threads implicitly created by the executor. This
+/// executor uses the scheduling hint to spawn threads with the first grouped on
+/// the first core, the second group getting the next consecutive threads, etc.
+/// For example, if 10 tasks are spawned (num_tasks is set to 10) and num_cores
+/// is set to 2 the executor will schedule the tasks in the following order:
+///
+/// worker thread | 1 | 2
+/// --------------+---+---
+/// tasks         | 1 | 6
+///               | 2 | 7
+///               | 3 | 8
+///               | 4 | 9
+///               | 5 | 10
+///
+/// rather than the typical round robin:
+///
+/// worker thread | 1 | 2
+/// --------------+---+---
+/// tasks         | 1 | 2
+///               | 3 | 4
+///               | 5 | 6
+///               | 7 | 8
+///               | 9 | 10
+///
+/// This executor is a one-way executor as apply/post is the only operation
+/// required by the HPX backend.
+struct ChunkedRoundRobinExecutor {
+  using execution_category = hpx::parallel::execution::parallel_execution_tag;
+
+  HPX_CONSTEXPR explicit ChunkedRoundRobinExecutor(
```

Comment:
Did you mean to decorate with `HPX_CONSTEXPR`?

Reply:
Yes. If you're wondering why it's `HPX_CONSTEXPR` and not `constexpr` it's because in HPX we've had cases where constexpr wasn't correctly supported in all compilers so we've put it behind a macro. That said, in this case it could quite safely just be `constexpr`.

Reply:
No other member function are declared `constexpr` so I did not see the point of declaring the constructor `constexpr`.

Reply:
Indeed. The `constexpr` isn't critical. I may have just put it there because I could.